nifi-nar-bundles/nifi-riemann-bundle/nifi-riemann-processors/src/main/java/org/apache/nifi/processors/riemann/PutRiemann.java

```diff
@@ -0,0 +1,376 @@
+    localProperties.add(ATTR_SERVICE);
+    localProperties.add(ATTR_STATE);
+    localProperties.add(ATTR_METRIC);
+    localProperties.add(ATTR_TTL);
+    localProperties.add(ATTR_TAGS);
+    localProperties.add(ATTR_HOST);
+    localProperties.add(ATTR_TIME);
+  }
+
+
+  @Override
+  public Set<Relationship> getRelationships() {
+    return relationships;
+  }
+
+  @Override
+  protected List<PropertyDescriptor> getSupportedPropertyDescriptors() {
+    return localProperties;
+  }
+
+  @Override
+  protected PropertyDescriptor getSupportedDynamicPropertyDescriptor(final String propertyDescriptorName) {
+    return new PropertyDescriptor.Builder()
+      .name(propertyDescriptorName)
+      .expressionLanguageSupported(true)
+      .addValidator(Validator.VALID)
+      .required(false)
+      .dynamic(true)
+      .build();
+  }
+
+  @OnStopped
+  public final void cleanUpClient() {
+    if (riemannClient != null) {
+      this.riemannClient.close();
+    }
+    this.riemannClient = null;
+    this.batchSize = -1;
+    this.customAttributes.clear();
+  }
+
+  @OnScheduled
+  public void onScheduled(ProcessContext context) throws ProcessException {
+    if (batchSize == -1) {
+      batchSize = context.getProperty(BATCH_SIZE).asInteger();
+    }
+    if (riemannClient == null || !riemannClient.isConnected()) {
+      transport = Transport.valueOf(context.getProperty(TRANSPORT_PROTOCOL).getValue());
+      String host = context.getProperty(RIEMANN_HOST).getValue().trim();
+      int port = context.getProperty(RIEMANN_PORT).asInteger();
+      writeTimeout = context.getProperty(TIMEOUT).asLong();
+      RiemannClient client = null;
+      try {
+        switch (transport) {
+          case TCP:
+            client = RiemannClient.tcp(host, port);
+            break;
+          case UDP:
+            client = RiemannClient.udp(host, port);
+            break;
+        }
+        client.connect();
+        riemannClient = client;
+      } catch (IOException e) {
+        context.yield();
+        throw new ProcessException(String.format("Unable to connect to Riemann [%s:%d] (%s)\n%s", host, port, transport, e.getMessage()));
+      }
+    }
+
+    if (customAttributes.size() == 0) {
+      for (Map.Entry<PropertyDescriptor, String> property : context.getProperties().entrySet()) {
+        // only custom defined properties
+        if (!getSupportedPropertyDescriptors().contains(property.getKey())) {
+          customAttributes.add(property.getKey());
+        }
+      }
+    }
+  }
+
+
+  @Override
+  public void onTrigger(ProcessContext context, ProcessSession session) throws ProcessException {
+    // Check if the client is currently connected, as a previous trigger could have detected a failure
+    // in the connection.
+    if (riemannClient == null || !riemannClient.isConnected()) {
+      // clean up the client and attempt to re-initialize the processor
+      cleanUpClient();
+      onScheduled(context);
```

Comment:
This block is what is spurring the leak when comms cannot be established to the remote system.  Is RiemannClient threadsafe?  This would be shared across multiple threads should concurrency which could cause issues given the logic currently in place.  If not, would likely prefer to see client generation per onTrigger.  Otherwise, need some better handling of the resource instantiation and teardown.